flink-tests/src/test/java/org/apache/flink/test/checkpointing/AutoRescalingITCase.java

```diff
@@ -0,0 +1,935 @@
+    }
+
+    @Test
+    public void testCheckpointRescalingOutKeyedState() throws Exception {
+        testCheckpointRescalingKeyedState(true);
+    }
+
+    /**
+     * Tests that a job with purely keyed state can be restarted from a savepoint with a different
+     * parallelism.
+     */
+    public void testCheckpointRescalingKeyedState(boolean scaleOut) throws Exception {
+        final int numberKeys = 42;
+        final int numberElements = 1000;
+        final int parallelism = scaleOut ? numSlots / 2 : numSlots;
+        final int parallelism2 = scaleOut ? numSlots : numSlots / 2;
+        final int maxParallelism = 13;
+
+        Duration timeout = Duration.ofMinutes(3);
+        Deadline deadline = Deadline.now().plus(timeout);
+
+        ClusterClient<?> client = cluster.getClusterClient();
+
+        try {
+
+            JobGraph jobGraph =
+                    createJobGraphWithKeyedState(
+                            new Configuration(),
+                            parallelism,
+                            maxParallelism,
+                            numberKeys,
+                            numberElements);
+
+            final JobID jobID = jobGraph.getJobID();
+
+            client.submitJob(jobGraph).get();
+
+            // wait til the sources have emitted numberElements for each key and completed a
+            // checkpoint
+            assertTrue(
+                    SubtaskIndexFlatMapper.workCompletedLatch.await(
+                            deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS));
+
+            // verify the current state
+
+            Set<Tuple2<Integer, Integer>> actualResult = CollectionSink.getElementsSet();
+
+            Set<Tuple2<Integer, Integer>> expectedResult = new HashSet<>();
+
+            for (int key = 0; key < numberKeys; key++) {
+                int keyGroupIndex = KeyGroupRangeAssignment.assignToKeyGroup(key, maxParallelism);
+
+                expectedResult.add(
+                        Tuple2.of(
+                                KeyGroupRangeAssignment.computeOperatorIndexForKeyGroup(
+                                        maxParallelism, parallelism, keyGroupIndex),
+                                numberElements * key));
+            }
+
+            assertEquals(expectedResult, actualResult);
+
+            // clear the CollectionSink set for the restarted job
+            CollectionSink.clearElementsSet();
+
+            waitForAllTaskRunning(cluster.getMiniCluster(), jobGraph.getJobID(), false);
+
+            waitForOneMoreCheckpoint(jobID, cluster.getMiniCluster());
+
+            JobResourceRequirements.Builder builder = JobResourceRequirements.newBuilder();
+            for (JobVertex vertex : jobGraph.getVertices()) {
+                builder.setParallelismForJobVertex(vertex.getID(), parallelism2, parallelism2);
+            }
+
+            restClusterClient.updateJobResourceRequirements(jobID, builder.build()).join();
```

Comment:
IIRC, this call doesn't wait for the requirements to be satisfied.
Should we wait after this call, that the actual DoP is actually what we want?

Reply:
I was looking into this and think it is not needed for this test because the test job will only start completing after it restarted from the rescale, so we can be sure that the rescaling happened first.

Reply:
Relying on that condition seems a bit fragile to me and makes it more difficult to understand.
- there were no restarts; they are disabled here, but I'm not sure if the restarts are disabled in every tests, and whether they will stay as such.
- the DoP might change if some TM is slow to start; the scheduler will first schedule with lower DoP, and once TM is up, will rescale to full DoP. That happened in some tests previously